src/lib/datatypes/include/sol-memdesc.h

```diff
@@ -0,0 +1,1089 @@
+ * @see sol_memdesc_free()
+ */
+int sol_memdesc_free_content(const struct sol_memdesc *desc, void *memory);
+
+/**
+ * @brief Free the contents and the memory.
+ *
+ * @param desc the memory description.
+ * @param memory the memory to free the contents and the memory itself.
+ *
+ * @see sol_memdesc_free_content()
+ */
+static inline void
+sol_memdesc_free(const struct sol_memdesc *desc, void *memory)
+{
+    sol_memdesc_free_content(desc, memory);
+    free(memory);
+}
+
+/**
+ * @brief Allocate the memory required by this description and initialize it.
+ *
+ * This will allocate offset + size bytes, then fill these bytes with
+ * the content defined in struct sol_memdesc::defcontent.
+ *
+ * @param desc the memory description.
+ *
+ * @return NULL on error, newly allocated memory on success. Free
+ * using sol_memdesc_free().
+ *
+ * @see sol_memdesc_free()
+ */
+static inline void *
+sol_memdesc_new_with_defaults(const struct sol_memdesc *desc)
+{
+    void *mem;
+    uint16_t size;
+    int r;
+
+    size = sol_memdesc_get_size(desc);
+    if (!size)
+        return NULL;
+
+    mem = malloc(size);
+    if (!mem)
+        return NULL;
+
+    r = sol_memdesc_init_defaults(desc, mem);
+    if (r < 0) {
+        sol_memdesc_free(desc, mem);
+        errno = -r;
+        return NULL;
+    }
+
+    errno = 0;
+    return mem;
+}
+
+/**
+ * @brief Get the length of an array.
+ *
+ * This function must be applied to SOL_MEMDESC_TYPE_ARRAY and will
+ * call struct sol_memdesc::ops::array::get_length.
+ *
+ * The returned value is about the number of items according to
+ * struct sol_memdesc::array_item.
+ *
+ * @param array_desc the memory description of type SOL_MEMDESC_TYPE_ARRAY.
+ * @param memory the memory holding the array.
+ *
+ * @return On error, negative errno is returned. Zero or more for success.
+ */
+ssize_t sol_memdesc_get_array_length(const struct sol_memdesc *array_desc, const void *memory);
+
+/**
+ * @brief Get the array element.
+ *
+ * This function must be applied to SOL_MEMDESC_TYPE_ARRAY and will
+ * call struct sol_memdesc::ops::array::get_element.
+ *
+ * @note for speed purposes, this function will not guarantee
+ * out-of-bounds checking, please ensure the index is less than
+ * sol_memdesc_get_array_length() before calling it.
+ *
+ * @param array_desc the memory description of type SOL_MEMDESC_TYPE_ARRAY.
+ * @param memory the memory holding the array.
+ * @param idx the index of the element inside the array.
+ *
+ * @return On error NULL is returned and errno is set. On success the
+ *         memory of the item is returned.
+ *
+ * @see sol_memdesc_get_array_length()
+ */
+void *sol_memdesc_get_array_element(const struct sol_memdesc *array_desc, const void *memory, size_t idx);
+
+/**
+ * @brief Resize the length of an array.
+ *
+ * This function must be applied to SOL_MEMDESC_TYPE_ARRAY and will
+ * call struct sol_memdesc::ops::array::resize.
+ *
+ * The returned value is about the number of items according to
```

Comment:
What 'returned value'? Return of this function? If so, this description doesn't match `@return` one.